### ISSUE TITLE
-b, --nocolor options not working usefully

### ISSUE THREAD

**Issue by mc-butler** [MidnightCommander/mc#3023]
> [!IMPORTANT]
> This issue was migrated from Trac:
>
> | **Origin** | https://midnight-commander.org/ticket/3023 |
> | :------------- | :------------------------------------------------------------------------------------------------ |
> | **Reporter** | `dnied` (dnied@….it) |

<p>
The '-b' and '--nocolor' mc options don't produce usable output because the selection bar is not visible. Mc should probably use inverse-screen for the selection bar when asked to go colorless.<br />
</p>
<p>
Experimenting with $TERM, I found that mc actually draws an inverse-screen selection bar, though only if TERM=xtermm (but then the arrow keys don't work). Mc should draw it like that for any value of TERM, when '-b' or '--nocolor' are specified.<br />
</p>
<p>
Verified on 'xterm -cm', with and without GNUscreen.<br />
</p>

```
~> mc -V
GNU Midnight Commander 4.8.4
Built with GLib 2.32.4
Using the S-Lang library with terminfo database
With builtin Editor
With subshell support as default
With support for background operations
With mouse support on xterm and Linux console
With support for X11 events
With internationalization support
With multiple codepages support
Virtual File Systems: cpiofs, tarfs, sfs, extfs, ftpfs, fish, smbfs
Data types: char: 8; int: 32; long: 32; void *: 32; size_t: 32; off_t: 64;
~> xterm -version
X.Org 7.6.0(281)
~>
```

**Comment by egmontkob** [comment]
I've just opened closely related #4985 to fix the engine, while keeping this bug for fixing the hardcoded black-and-white skin.

Which one is the _intent_ of black-and-white mode?
- Build from the four colors "black" (index 0), "brightblack" (a.k.a. "gray", index 8), "white" (a.k.a. "lightgray", index 7) and "brightwhite" (a.k.a. "white", index 15)?
- Don't emit any color changing instruction whatsoever, i.e. use 2 colors only, whatever the terminal's default colors are; combined with "reverse" and other attributes?

Currently mc implements the first one (4 colors) if compiled against slang, and the second one (2 colors) if compiled against ncurses. (The latter one I don't understand sourcecode-wise yet.)

**Comment by egmontkob** [comment]
Looks like it is intentional that mc tries to (and less intentional that it fails to) implement both a colorless (2-color) and a 4-color scheme, with both screen libraries. `mc_tty_color_pair_init_special()` receives two sets of colors, and picks one based on whether in 2-color or 4-color mode.

I'm tempted to say that we should simplify our lives as much as possible, i.e. only have a hardwired mode for the last resort fallback 2-color look. A 4-color look, if required, can be implemented as a regular skin.

A tiny problem with that is that editor syntax highlighting colors are hardcoded and not skinnable, there it'd shine in many colors, but that can be addressed differently if we ever get there (e.g. a keyword in the skin file to disable syntax highlighting).

Unless someone objects, I'll go with the truly monochrome / colorless / 2-color (call it whatever you want) mode.

**Comment by egmontkob** [comment]
While at it, I'll stop using `underline` decoration for anything in the hardcoded b&w skin. I'll only use `bold` and `reverse` (and the combination of these two).

It's a last resort fallback skin. If a terminal doesn't support colors then almost certainly it doesn't support underline either.

If the terminal does support all these fancy features, it's just the user's preference not to use most of them, then, again, this is what skin support is for.

**Comment by ossilator** [comment]
> It's a last resort fallback skin.

what for? mc doesn't have to be failsafe in a corrupted installation.

> If a terminal doesn't support colors then almost certainly it doesn't support underline either.

that's just wrong.
i actually used to own [a monochrome terminal](https://www.vecmar.com/pdf/wyse/WY-99GT.pdf). it had all kinds of fancy features, including graphics mode.
the same is true for the hercules card (and the ibm mda it was derived from, without graphics).

whether such things should be still considered relevant is debatable.

**Comment by egmontkob** [comment]
> what for? mc doesn't have to be failsafe in a corrupted installation.

Not a corrupted installation, but a super limited terminal.

I wouldn't mind completely removing the black-and-white hardwired skin and the `--color` and `--nocolor` options. Maybe the 4.9 release is the right time for that?  I just didn't want to go this far with this change, as I'm not familiar with such setups where people might be looking for this option.

Are we okay to require at least 8 colors (and a corresponding `$TERM`)? @zyv @mc-worker What do you think?

Personally, while I generally love cleaning up legacy stuff, this one doesn't bother me, I'm fine either way.

Now that I've invested the work to fix its problems, maybe let that change land first, and then we can remove the option (just in case we'd need to resurrect it, we'd get the fixed version).

> that's just wrong.

I didn't say it certainly doesn't support, I said it likely doesn't. I might be wrong as I'm not well-versed with legacy hardware terminals.

`TERM=xtermm` only describes bold and reverse as supported, not underline. The Linux console also only supports bold (brirght) and reverse, not underline.

For a last resort skin, I wouln't want to rely on features that might easily be unavailable, I want to target the bare common minimum, to be sure that the skin is usable even then.

**Comment by egmontkob** [comment]
> It's a last resort fallback skin.

Just for the record:

I don't _know_ for a fact that it's a last resort fallback. I don't know the original intent.

I'm reverse engineering that I think it _must be_ intended as a last resort option.

Because the other choice is that the user prefers to have such a pure look; if so, that should be achieved via an external skin.

If we remove the hardwired option, I think we should add such an external skin.

**Comment by egmontkob** [comment]
Oh, one more important thing.

> mc doesn't have to be failsafe in a corrupted installation.

I'm not 100% on this.

A possible "corrupted" installation is if mc's external files are completely missing. Maybe you have just "make"d the binary but you haven't run "make install" yet, or the skins are missing for some other reason.

If a complex projects needs dozens or hundreds of external files to even provide minimun functionality then that's fine. If mc already required a bunch of such external files I wouldn't complain.

But I don't want the graphical skin to be the first one.

If the removal of the `--nocolor` option and the corresponding internal basic skin would mean that `mc` cannot any more start up after being built, without having "make install"ed its external files, then I'd rather keep this legacy feature and keep `mc` usable as a standalone binary.

Strictly speaking, it is indeed a "corrupted" installation that doesn't have to work, but it is definitely practical (at the very least for me while developing mc) that it works. I'd prefer to keep this practical behavior.

**Comment by zyv** [comment]
> Are we okay to require at least 8 colors (and a corresponding `$TERM`)?

I would leave it, because it appears to me that removing it if it's working correctly is hardly a maintenance win. What I would rather like to remove would be stuff like hardcoded FISH fallback scripts...

**Comment by ossilator** [comment]
i agree that completely removing monochrome support "just because" seems a bit pointless.
however, relying on skins for it to avoid special-cased code seems perfectly reasonable. it should be integrated into the automatic selection/fallback mechanism. #4537 is somewhat related to that.

**Comment by egmontkob** [comment]
Currently the code invokes a few internal methods to set the hardwired bw skin. This has side effects, as noted at https://github.com/MidnightCommander/mc/issues/4981#issuecomment-3837448385.

We might instead hardwire the entire skin in the source, as a multi-line string constant, and pass it to the same code to process that processes external skins. This would probably simplify the code and its subsequent maintenance.

We could hardcode the bw skin like this, or the default, or both.

It would be great if the user could easily modify or fork this/these skin(s), without having to start from scratch or having to look up the skin's definition in mc's code.

This would mean that the chosen fallback skin(s) (either "bw" or "default" or both) would need to be duplicated: hardwired but also still shipped as an external skin. Strongly preferably without manually keeping them in sync. We'd need some magic (a build step) that converts this/these skin(s) from the `.ini` file to a source code fragment. It would be a one-off solution (plenty of people have already done it, there are tons of webfinds). `#embed` is pretty new (C23) so I guess we should refrain from using it for many years to come, no matter how convenient it would be.
